libsolidity/ast/Types.cpp

```diff
@@ -704,6 +704,33 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 				value = rational(denominator, numerator);
 			break;
 		}
+		case Token::SHL:
+		{
+			using boost::multiprecision::pow;
+			if (fractional)
+				return TypePointer();
+			else if (other.m_value < 0)
+				return TypePointer();
+			else if (other.m_value > numeric_limits<uint32_t>::max())
+				return TypePointer();
+			uint32_t exponent = other.m_value.numerator().convert_to<uint32_t>();
+			value = m_value.numerator() * pow(bigint(2), exponent);
+			break;
+		}
+		// NOTE: we're using >> (SAR) to denote (SHR)
```

Comment:
In the end `SAR` is implemented, because `-0x4200 >> 8` keeps the sign.  Maybe this comment should be
```
NOTE: we're using >> (SAR) to denote (SAR)
```
or not needed.